File: cpp/0974-subarray-sums-divisible-by-k.cpp
#include <unordered_map>
#include <vector>

class Solution {
    public:
        [[nodiscard]] auto
        subarraysDivByK(const std::vector<int> &nums, const int k) -> int {
            auto prefix_sum = 0;
            // auto freqs = std::unordered_map<int, int>{{0, 1}};
            auto freqs = std::vector<int>(static_cast<std::size_t>(k));
            freqs[0] = 1;
            auto ans = 0;
            for (const auto n : nums) {
                prefix_sum = ((prefix_sum + n) % k + k) % k;
                const auto i = static_cast<std::size_t>(prefix_sum);
                ans += freqs[i];
                ++freqs[i];
            }

            return ans;
        }
};

File: py/0974-subarray-sums-divisible-by-k.py
class Solution:
    def subarraysDivByK(self, nums: list[int], k: int) -> int:
        prefix_sum = 0
        # freqs = defaultdict(int)
        freqs = [0] * k
        freqs[0] = 1
        ans = 0
        for n in nums:
            prefix_sum = (prefix_sum + n) % k
            ans += freqs[prefix_sum]
            freqs[prefix_sum] += 1

        return ans
